lib/rust/prelude/src/range.rs

```diff
@@ -0,0 +1,211 @@
+// // ================
+//
+// pub trait RangeLen: RangeOps {
+//     fn len(&self) -> <Self::Item as std::ops::Sub<Self::Item>>::Output
+//     where Self::Item: std::ops::Sub<Self::Item>;
+// }
+//
+// impl<T: Clone> RangeLen for std::ops::Range<T> {
+//     fn len(&self) -> <Self::Item as std::ops::Sub<Self::Item>>::Output
+//     where Self::Item: std::ops::Sub<Self::Item> {
+//         self.end.clone() - self.start.clone()
+//     }
+// }
+//
+// impl<T: Clone + std::ops::Sub> RangeLen for std::ops::RangeInclusive<T>
+// where <T as std::ops::Sub>::Output: std::iter::Step
+// {
+//     fn len(&self) -> <Self::Item as std::ops::Sub<Self::Item>>::Output
+//     where Self::Item: std::ops::Sub<Self::Item> {
+//         <<T as std::ops::Sub>::Output as std::iter::Step>::forward(
+//             self.end().clone() - self.start().clone(),
+//             1,
+//         )
+//     }
+// }
+
+
+
+// ====================
+// === RangeOverlap ===
+// ====================
+
+pub trait RangeOverlap<T = Self> {
+    fn overlaps(&self, other: &T) -> bool;
+}
+
+impl<T: PartialOrd> RangeOverlap<std::ops::Range<T>> for std::ops::Range<T> {
+    fn overlaps(&self, other: &Self) -> bool {
+        let overlap1 = || self.start >= other.start && self.start < other.end;
+        let overlap2 = || other.start >= self.start && other.start < self.end;
+        overlap1() || overlap2()
+    }
+}
+
+impl<T: PartialOrd> RangeOverlap<std::ops::RangeInclusive<T>> for std::ops::RangeInclusive<T> {
+    fn overlaps(&self, other: &Self) -> bool {
+        let overlap1 = || self.start() >= other.start() && self.start() <= other.end();
+        let overlap2 = || other.start() >= self.start() && other.start() <= self.end();
+        overlap1() || overlap2()
+    }
+}
+
+
+// ======================
+// === RangeIntersect ===
+// ======================
+
+pub trait RangeIntersect<T = Self> {
+    type Result;
+    fn intersect(&self, other: &T) -> Self::Result;
+}
+
+impl<T> RangeIntersect<std::ops::RangeInclusive<T>> for std::ops::RangeInclusive<T>
+where T: PartialOrd + Clone
+{
+    type Result = Option<std::ops::RangeInclusive<T>>;
+
+    fn intersect(&self, other: &Self) -> Self::Result {
+        if self.overlaps(other) {
+            let self_start = self.start().clone();
+            let self_end = self.end().clone();
+            let other_start = other.start().clone();
+            let other_end = other.end().clone();
+            let start = if self_start > other_start { self_start } else { other_start };
+            let end = if self_end < other_end { self_end } else { other_end };
+            Some(start..=end)
+        } else {
+            None
+        }
+    }
+}
+
+
+
+// ================================
+// === merge_overlapping_ranges ===
+// ================================
+
+pub fn merge_overlapping_ranges<R>(ranges: &[R]) -> impl Iterator<Item = R>
+where
+    R: Clone + RangeOverlap + RangeOps,
+    <R as RangeOps>::Item: Clone + PartialOrd, {
```

Comment:
Is there a need to provide this API when `Item` is only `PartialOrd`? This implementation is only correct if the inputs received at runtime are in practice totally-ordered.

Reply:
Good catch, what would you propose here?

Reply:
Tighten the `PartialOrd` bound to `Ord`, and sort with plain `sort_unstable()`.

Also, an optimization: Since the function needs its ranges in a Vec unconditionally, it could accept its input as a Vec and save an allocation.

Reply:
I applied all the suggestions but the one with using `sort_unstable` instead of `sort_unstable_by`. `std::ops::Range` does not implement `Ord`, so its impossible.